### PR TITLE
feat: add `enarx sign` and `--signatures` option

### PR DESCRIPTION
**Pull request by haraldh**
Fixes: https://github.com/enarx/enarx/issues/2073

To run signed keeps a signature file has to be created first:

```console
$ openssl genrsa -3 -out sgx.key 3072
$ openssl genpkey -outform PEM -out sev-id.key -algorithm EC -pkeyopt ec_paramgen_curve:P-384 -pkeyopt ec_param_enc:named_curve
$ openssl genpkey -outform PEM -out sev-author.key -algorithm EC -pkeyopt ec_paramgen_curve:P-384 -pkeyopt ec_param_enc:named_curve
```

```console
$ enarx sign --sgx-key sgx.key --sev-key sev-id.key --sev-author-key sev-author.key  > sig.json
```

Then to run a workload the `--signatures` option can be used to start signed Keeps:
```console
$ enarx run --signatures sig.json hello_wasi_snapshot1.wasm
```

Signing and running with the signatures is now enabled in the wasm test suite!

Signed-off-by: Harald Hoyer <harald@profian.com>

### CODE REVIEW

File: src/backend/sev/builder.rs [hunk 1]
```diff
@@ -189,27 +185,53 @@ impl super::super::Mapper for Builder {
...
+            if sig_blob.id_auth.len() != (core::mem::size_of::<IdAuth>()) {
+                return Err(anyhow!("Invalid SEV signature blob size."));
+            }
+
+            if sig_blob.id_block.len() != (core::mem::size_of::<IdBlock>()) {
+                return Err(anyhow!("Invalid SEV signature blob size."));
+            }
 
-        let finish = Finish::new(None, None, [0u8; 32]);
+            id_auth = unsafe { (sig_blob.id_auth.as_ptr() as *const IdAuth).read_unaligned() };
```
Comment (bstrie) [inline src/backend/sev/builder.rs:216]: Safety comments.

File: src/backend/sev/snp/launch/linux.rs [hunk 2]
```diff
@@ -224,26 +225,26 @@ pub struct LaunchFinish<'a> {
...
+            id_block_en: finish.id_block_n_auth.is_some().into(),
+            auth_key_en: finish.auth_key_en.into(),
             host_data: finish.host_data,
             pad: [0u8; 6],
             _phantom: PhantomData,
         }
     }
 }
+
+pub const SEV_SNP_VMSA_SHA384: [u8; 48] = [
```
Comment (bstrie) [inline src/backend/sev/snp/launch/linux.rs:246]: Where does this come from?

File: crates/shim-kvm/src/snp/launch.rs [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+        let flags = self.flags.bits() | 0b10;
+        let mut flags_64 = flags as u64;
+
+        major_version <<= 8;
+        flags_64 <<= 16;
+
+        val |= minor_version;
+        val |= major_version;
+        val |= flags_64;
+        val &= 0x00FFFFFF;
```
[the remark this answers is not in the document]
Comment (haraldh) [reply crates/shim-kvm/src/snp/launch.rs:70]: https://github.com/enarx/sev/blob/main/src/launch/snp.rs#L161
masks all the other invalid bits per SEV-SNP firmware spec

I'll add a comment

File: src/backend/sgx/mod.rs [hunk 4]
```diff
@@ -62,13 +63,18 @@ impl crate::backend::Backend for Backend {
...
+        exec: &[u8],
+        signatures: Option<Signatures>,
+    ) -> Result<Arc<dyn super::Keep>> {
+        builder::Builder::load(shim, exec, signatures)
     }
 
     #[inline]
     fn hash(&self, shim: &[u8], exec: &[u8]) -> Result<Vec<u8>> {
-        hasher::Hasher::load(shim, exec)
+        hasher::Hasher::load(shim, exec, None)
```
[the remark this answers is not in the document]
Comment (haraldh) [reply src/backend/sgx/mod.rs:77]: in binary.rs
```
impl<T: Mapper> Loader for T {
    fn load(...)
```

[the remark this answers is not in the document]
Comment (haraldh) [reply src/backend/sgx/mod.rs:77]: I don't want to duplicate that code.

File: src/backend/sgx/builder.rs [hunk 5]
```diff
@@ -130,27 +130,35 @@ impl TryFrom<Builder> for Arc<dyn super::super::Keep> {
...
+            }
+            unsafe { (sig_blob.as_ptr() as *const Signature).read_unaligned() }
         } else {
-            RS256PrivateKey::generate(3).context("Failed to create RSA key")?
-        };
+            // Create the enclave signature
+            let hash = builder.hash.finish();
+            let author = Author::new(0, 0);
+            let body = builder.cnfg.parameters.body(hash);
+            let key = if let Ok(key) = std::env::var("ENARX_TEST_SGX_KEY_FILE")
```
[the remark this answers is not in the document]
Comment (haraldh) [reply src/backend/sgx/builder.rs:145]: Hmm, in production runs we always have a signature and therefore don't reach this branch.

File: src/cli/sign.rs [hunk 6]
```diff
@@ -0,0 +1,529 @@
...
+use p384::ecdsa::SigningKey;
+use p384::elliptic_curve::sec1::Coordinates;
+use p384::pkcs8::DecodePrivateKey;
+use p384::EncodedPoint;
+use sgx::crypto::{rcrypto::*, *};
+use sgx::signature::{Author, Body, Signature};
+
+impl ByteSized for ::sgx::signature::Signature {}
+
+/// Run a WebAssembly module inside an Enarx Keep.
```
[the remark this answers is not in the document]
Comment (haraldh) [reply src/cli/sign.rs:27]: fixed

File: src/cli/sign.rs [hunk 7]
```diff
@@ -0,0 +1,529 @@
...
+#[derive(Args, Debug)]
+pub struct Options {
+    /// Binary to load and run inside the keep
```
[the remark this answers is not in the document]
Comment (haraldh) [reply src/cli/sign.rs:30]: fixed

File: src/cli/sign.rs [hunk 8]
```diff
@@ -0,0 +1,529 @@
...
+    /// SEV P-384 private key in PEM form
+    #[clap(long)]
+    sev_key: Utf8PathBuf,
+}
+
+fn sign_sgx(body_bytes: &[u8], sgx_key: &RS256PrivateKey) -> Result<Vec<u8>> {
+    if body_bytes.len() != size_of::<Body>() {
+        bail!("Invalid length of SGX input data");
+    }
+    let body: Body = unsafe { (body_bytes.as_ptr() as *const Body).read_unaligned() };
```
[the remark this answers is not in the document]
Comment (haraldh) [reply src/cli/sign.rs:51]: fixed

File: src/cli/sign.rs [hunk 9]
```diff
@@ -0,0 +1,529 @@
...
+
+    // FIXME: do we need the date and sw defined value?
+    let author = Author::new(0, 0);
+    let sig = Signature::new(sgx_key, author, body).context("Failed to create RSA signature")?;
+
+    Ok(sig.as_bytes().to_vec())
+}
+
+fn sign_sev(
```
[the remark this answers is not in the document]
Comment (haraldh) [reply src/cli/sign.rs:60]: fixed

File: src/backend/sev/snp/mod.rs [hunk 10]
```diff
@@ -32,11 +32,30 @@ macro_rules! impl_const_id {
...
 
 use std::fmt::{Debug, Display, Formatter};
+use std::mem::size_of;
 use std::{error, io};
 
+pub trait ByteSized: Sized {
+    /// The constant default value.
+    const SIZE: usize = size_of::<Self>();
+
+    unsafe fn from_bytes(bytes: &[u8]) -> Option<Self> {
```
[the remark this answers is not in the document]
Comment (haraldh) [reply src/backend/sev/snp/mod.rs:46]: done

File: src/backend/sev/snp/mod.rs [hunk 11]
```diff
@@ -32,11 +32,30 @@ macro_rules! impl_const_id {
...
+        if bytes.len() != Self::SIZE {
+            return None;
+        }
+
+        Some((bytes.as_ptr() as *const _ as *const Self).read_unaligned())
+    }
+
+    fn as_bytes(&self) -> &[u8] {
+        unsafe { core::slice::from_raw_parts(self as *const _ as *const u8, Self::SIZE) }
```
[the remark this answers is not in the document]
Comment (haraldh) [reply src/backend/sev/snp/mod.rs:55]: done

File: src/backend/sev/builder.rs (hunk 1 above)
Comment (haraldh) [reply src/backend/sev/builder.rs]: done

File: src/backend/sev/snp/sign.rs [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+            s: [0u8; 0x48],
+        }
+    }
+}
+
+#[repr(C)]
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub struct Signature {
+    pub component: Es384,
+    rsvd: [u8; 368], // must be zero
```
[the remark this answers is not in the document]
Comment (haraldh) [reply src/backend/sev/snp/sign.rs]: done

File: src/backend/sev/snp/sign.rs [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+}
+
+impl ByteSized for Signature {}
+
+#[repr(C)]
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub struct PublicKey {
+    pub curve: u32,
+    pub component: Es384,
+    rsvd: [u8; 880], // must be zero
```
[the remark this answers is not in the document]
Comment (haraldh) [reply src/backend/sev/snp/sign.rs]: done

File: src/backend/sev/snp/launch/linux.rs (hunk 2 above)
Comment (haraldh) [reply src/backend/sev/snp/launch/linux.rs:254]: added comment

File: src/backend/sgx/builder.rs [hunk 14]
```diff
@@ -130,27 +130,35 @@ impl TryFrom<Builder> for Arc<dyn super::super::Keep> {
...
-            .and_then(|keystr| RS256PrivateKey::from_pem(&keystr).map_err(Error::new))
-        {
-            key
+        let signature = if let Some(signatures) = builder.cnfg.signatures {
+            let sig_blob = signatures.sgx;
+
+            if sig_blob.len() != core::mem::size_of::<Signature>() {
+                bail!("Invalid SGX signature length");
+            }
+            unsafe { (sig_blob.as_ptr() as *const Signature).read_unaligned() }
```
[the remark this answers is not in the document]
Comment (haraldh) [reply src/backend/sgx/builder.rs]: done

File: src/backend/sgx/hasher.rs [hunk 15]
```diff
@@ -38,6 +49,14 @@ impl TryFrom<Hasher> for Vec<u8> {
 
     #[inline]
     fn try_from(hasher: Hasher) -> Result<Self> {
-        Ok(hasher.0.finish().into())
+        let body = hasher.cnfg.parameters.body(hasher.digest.finish());
+
+        Ok(unsafe {
```
[the remark this answers is not in the document]
Comment (haraldh) [reply src/backend/sgx/hasher.rs:55]: done

**Review by bstrie — APPROVED** [review]

**Pull request merged**